sys/gen/boot.c

```diff
@@ -0,0 +1,60 @@
+#include <sys/boot.h>
+#include <libfdt/libfdt.h>
+
+/* Last physical address used by kernel for boot memory allocation. */
+__boot_data void *_bootmem_end;
+
+__boot_data static uintptr_t _boot_sbrk;
+
+/* Initialize boot memory allocator. */
+__boot_text void boot_sbrk_init(paddr_t ebss_pa) {
+  _boot_sbrk = align(ebss_pa, sizeof(long));
+}
+
+/* Clears and allocates clears physical memory just after kernel image end. */
+__boot_text void *boot_sbrk(size_t n) {
+  long *addr = (long *)_boot_sbrk;
+  _boot_sbrk += align(n, sizeof(long));
+  for (unsigned i = 0; i < n / sizeof(long); i++)
+    addr[i] = 0;
+  return addr;
+}
+
+__boot_text paddr_t boot_sbrk_align(size_t n) {
+  _boot_sbrk = align(_boot_sbrk, n);
+  return (paddr_t)_boot_sbrk;
+}
+
+__boot_text void boot_clear(paddr_t start, paddr_t end) {
+  long *s = (long *)start;
+  long *e = (long *)end;
+  for (; s < e; s++)
+    *s = 0;
+}
+
+__boot_text static inline uint32_t fdt32toh(fdt32_t u) {
```

Comment:
If `bswap32` is not supported directly by the processor it will jump into library code stored in virtual address space. I defined the function specifically to avoid this problem.

Reply:
Added a comment.